pyomo/contrib/gdpopt/util.py

```diff
@@ -187,14 +187,89 @@ def fix_discrete_var(var, val, config):
         else:
             var.fix(val, skip_validation=True)
 
-class fix_discrete_problem_solution_in_subproblem(object):
+
+class fix_discrete_solution_in_subproblem(object):
+    def __init__(self, true_disjuncts, boolean_var_values, integer_var_values,
+                 subprob_util_block, config, solver):
+        self.True_disjuncts = true_disjuncts
+        self.boolean_var_values = boolean_var_values
+        self.discrete_var_values = integer_var_values
+        self.subprob_util_block = subprob_util_block
+        self.config = config
+
+    def __enter__(self):
+        # fix subproblem Blocks according to the discrete problem solution
+        fixed = []
+        for block in self.subprob_util_block.disjunct_list:
+            if block in self.True_disjuncts:
+                block.binary_indicator_var.fix(1)
+                fixed.append(block.name)
+            else:
+                block.deactivate()
+                block.binary_indicator_var.fix(0)
+        self.config.logger.debug("Fixed the following Disjuncts to 'True': %s"
+                            % ", ".join(fixed))
+
+        fixed_bools = []
+        for subprob_bool, val in zip(
+                self.subprob_util_block.non_indicator_boolean_variable_list,
+                self.boolean_var_values):
+            subprob_binary = subprob_bool.get_associated_binary()
+            if val:
+                subprob_binary.fix(1)
+            else:
+                subprob_binary.fix(0)
+            fixed_bools.append("%s = %s" % (subprob_bool.name, val))
+        self.config.logger.debug("Fixed the following Boolean variables: %s"
+                            % ", ".join(fixed_bools))
+
+        # Fix subproblem discrete variables according to the discrete problem
+        # solution
+        if self.config.force_subproblem_nlp:
+            fixed_discrete = []
+            for subprob_var, val in zip(
+                    self.subprob_util_block.discrete_variable_list,
+                    self.discrete_var_values):
+                fix_discrete_var(subprob_var, val, self.config)
+                fixed_discrete.append("%s = %s" % (subprob_var.name, val))
+            self.config.logger.debug("Fixed the following integer variables: "
+                                     "%s" % ", ".join(fixed_discrete))
+
+        # Call the subproblem initialization callback
+        self.config.subproblem_initialization_method(self.True_disjuncts,
+                                                     self.boolean_var_values,
+                                                     self.discrete_var_values,
+                                                     self.subprob_util_block)
```

Comment:
As a general matter of practice, `__enter__()` should return something (usually `self`)

Reply:
Oh, I did not know this. I just made these two return themselves, and also the suppress-infeasibility-warning one.